### PR TITLE
Add an admin form to pre-populate an LDAP user

### PR DESCRIPTION
**Pull request by scholtalbers**
Includes PR #5136 and that could therefore be closed.

With the added galaxy option `show_prepopulate_form` the following form behavior is shown:

screen shot 2017-12-20 at 21 07 44

⬇️

screen shot 2017-12-20 at 21 09 41

⬇️

screen shot 2017-12-20 at 21 51 01

### CODE REVIEW

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by scholtalbers** [comment]
@martenson anything I can do to speed up a review/possible merge - i.e. are there any major things that should be changed to consider this and the related PR?

**Comment by martenson** [comment]
@scholtalbers It seems fine to me but I do not know much about LDAP. It is holiday season so it might take some time to get this merged. Are you trying to get this into 18.01?

**Comment by scholtalbers** [comment]
I would like to get it in as quick as possible 😉 - to avoid having my own production instance code base too diverged from main. But I am aware of the holiday season🎅🏿 and due to the changes to the login system it may require some thorough review. Basically wanted to know if anything obvious was wrong with this  🙂

**Comment by martenson** [comment]
@scholtalbers Are you running dev in production? I cannot guarantee it will get to 18.01 but I will try to help making it happen.

**Comment by scholtalbers** [comment]
No, running 17.09 at the moment actually - but things can change..
Thanks @martenson!

**Comment by scholtalbers** [comment]
Maybe this can help a reviewer: https://github.com/scholtalbers/docker-galaxy-openldap
No automated test, and not ideal (e.g. the `.ldif` can certainly be extended), but its a start.

File: config/galaxy.ini.sample [hunk 1]
```diff
@@ -1045,6 +1045,10 @@ use_interactive = True
 # Allow administrators to log in as other users (useful for debugging)
 #allow_user_impersonation = False
 
+# When using LDAP for authentication, allow administrators to pre-populate users
+# using an additional form on 'Create new user'
+#show_prepopulate_form = False
```
Comment (natefoo) [inline config/galaxy.ini.sample]: Maybe `show_user_prepopulate_form` just to make it a bit more domain specific.

File: lib/galaxy/auth/util.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        authenticator = Authenticator(
+            plugin=plugin,
+            filter_template=filter_template,
+            options=options,
+        )
+        authenticators.append(authenticator)
+    return authenticators
+
+
+def parse_auth_results(trans, auth_results, options):
```
Comment (natefoo) [inline lib/galaxy/auth/util.py:50]: Moving this to a function I guess allays some of my concerns about this data structure raised in #5136

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 3]
```diff
@@ -150,13 +171,53 @@ def authenticate(self, email, username, password, options):
...
+        return (True,
+                _get_subs(options, 'auto-register-email', params),
+                _get_subs(options, 'auto-register-username', params),
+                attributes)
+
+    def _authenticate(self, params, options):
+        """
+        Do the actual authentication by binding as the user to check their credentials
+        """
+        import ldap
```
Comment (natefoo) [inline lib/galaxy/auth/providers/ldap_ad.py:221]: Typically we put this at the top with the regular imports:

```python
try:
    import ldap
except ImportError as exc:
    ldap = None
    ldap_import_exc = exc
```

And then here you'd do something like:

```python
if ldap is None:
    raise RuntimeError("Failed to load LDAP module: %s", str(ldap_import_exc))
```

**Review by natefoo — COMMENTED** [review]
Thanks @scholtalbers! See also my comments in #5136, which I guess we can close in favor of this one.

**Comment by scholtalbers** [comment]
Thanks for the comments @natefoo - hope I mitigated (most of) your concerns.

**Comment by natefoo** [comment]
Thanks! I went ahead and did the merge to resolve the conflict since that was essentially my doing in the first place.

+1 from me, just going to wait for tests.

**Comment by natefoo** [comment]
@galaxybot test this

**Comment by mvdbeek** [comment]
The failing tests are unrelated and have been fixed in dev, so I'll go ahead and merge this. Thanks a lot @scholtalbers, this is pretty cool. We should make use of the docker container for some integration tests in the future.

File: lib/tool_shed/util/admin_util.py [hunk 4]
```diff
@@ -89,27 +89,8 @@ def create_role(self, trans, **kwd):
...
-                    group = trans.app.model.Group(name=name)
-                    trans.sa_session.add(group)
-                    # Associate the group with the role
-                    gra = trans.model.GroupRoleAssociation(group, role)
-                    trans.sa_session.add(gra)
-                    num_in_groups = len(in_groups) + 1
-                else:
-                    num_in_groups = len(in_groups)
-                trans.sa_session.flush()
+                role, num_in_groups = trans.app.security_agent.create_role(
```
Comment (martenson) [inline lib/tool_shed/util/admin_util.py:92]: Toolshed does not have this method in its security agent (which is different than Galaxy's). Should be fixed in https://github.com/galaxyproject/galaxy/pull/6026/commits/40d0d997d38ea6832c83a78c880db0af5c23a90a

@scholtalbers

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 5]
```diff
@@ -140,23 +162,63 @@ def authenticate(self, email, username, password, options):
...
+                        elif attr in attrs:
                             params[attr] = str(attrs[attr][0])
                         else:
                             params[attr] = ""
+
+                if self.auto_create_roles_or_groups and self.role_search_option not in params:
+                    raise ConfigurationError("Missing or mismatching LDAP parameters for %s. Make sure the %s is "
+                                             "included in the 'search-fields'." %
+                                             (self.role_search_option, self.role_search_attribute))
+                log.critical(params)
```
Comment (nsoranzo) [inline lib/galaxy/auth/providers/ldap_ad.py:183]: @scholtalbers I guess this was a debugging statement? I think it should be changed to `log.debug()` or removed?

Comment (scholtalbers) [reply lib/galaxy/auth/providers/ldap_ad.py:183]: seems like it, can be removed.

Comment (nsoranzo) [reply lib/galaxy/auth/providers/ldap_ad.py:183]: Cool, I'll do that.

**Pull request merged**